src/Swarm/Game/ScenarioStatus.hs

```diff
@@ -0,0 +1,319 @@
+{-# LANGUAGE TemplateHaskell #-}
+{-# OPTIONS_GHC -Wno-orphans #-}
+{-# OPTIONS_GHC -Wno-partial-fields #-}
+
+-- -Wno-orphans is for the Eq/Ord Time instances
+
+-- |
+-- Module      :  Swarm.Game.ScenarioStatus
+-- Copyright   :  Brent Yorgey
+-- Maintainer  :  byorgey@gmail.com
+--
+-- SPDX-License-Identifier: BSD-3-Clause
+--
+-- Code to save and restore the status of scenarios, i.e. which have been
+-- started, which have been completed, etc.
+module Swarm.Game.ScenarioStatus (
+  ScenarioStatus (..),
+  _NotStarted,
+  _InProgress,
+  _Complete,
+  ScenarioInfo,
+  scenarioPath,
+  scenarioStatus,
+  scenarioBest,
+  updateScenarioInfoOnQuit,
+
+  -- * Scenario collection
+  ScenarioCollection (..),
+  scenarioCollectionToList,
+  scenarioItemByPath,
+  normalizeScenarioPath,
+  ScenarioItem (..),
+  scenarioItemName,
+  _SISingle,
+
+  -- * Loading and saving scenarios
+  loadScenarios,
+  loadScenarioInfo,
+  saveScenarioInfo,
+
+  -- * Re-exports
+  module Swarm.Game.Scenario,
+) where
+
+import Control.Algebra (Has)
+import Control.Carrier.Lift (Lift, sendIO)
+import Control.Carrier.Throw.Either (Throw, runThrow, throwError)
+import Control.Lens hiding (from, (<.>))
+import Control.Monad (unless, when)
+import Data.Aeson (
+  Options (..),
+  defaultOptions,
+  genericParseJSON,
+  genericToEncoding,
+  genericToJSON,
+ )
+import Data.Char (isSpace, toLower)
+import Data.Function (on)
+import Data.List (intercalate, stripPrefix, (\\))
+import Data.Map (Map)
+import Data.Map qualified as M
+import Data.Maybe (isJust)
+import Data.Text (Text, pack)
+import Data.Time (NominalDiffTime, ZonedTime, diffUTCTime, zonedTimeToUTC)
+import Data.Yaml as Y
+import GHC.Generics (Generic)
+import Paths_swarm (getDataDir)
+import Swarm.Game.Entity
+import Swarm.Game.Scenario
+import Swarm.Util (getSwarmSavePath)
+import System.Directory (canonicalizePath, doesDirectoryExist, doesFileExist, listDirectory)
+import System.FilePath (pathSeparator, splitDirectories, takeBaseName, takeExtensions, (-<.>), (</>))
+import Witch (into)
+
+instance Eq ZonedTime where
+  (==) = (==) `on` zonedTimeToUTC
+
+instance Ord ZonedTime where
+  (<=) = (<=) `on` zonedTimeToUTC
+
+data ScenarioStatus
+  = NotStarted
+  | InProgress
```

Comment:
Since we can't save games yet (#50), I guess "in progress" at this point really means "you tried it but gave up"?  In that case I don't know why we really need `_scenarioElapsed`.

This whole business with keeping track of elapsed time is tricky, and I don't necessarily have deeply thought reasons for my use of `ZonedTime` and `NominalDiffTime`.  In general, there are a few things at play here.  First, according to the reasoning at https://github.com/swarm-game/swarm/wiki/Speedrunning , I think the way to measure speedruns is to measure wall clock time and disallow adjusting the ticks/second rate.   However, when deciding how many ticks to run each frame etc., we are very careful to use a monotonic clock so that e.g. the game does not freeze while suddenly trying to run an hour's worth of ticks when the time zone changes or whatever.  So simply using `ZonedTime` is actually a bit problematic in case your game takes place across some kind of clock shift.  Of course this is not going to be common, but if we can solve the problem properly, then why not.  A third issue is that people are going to be used to seeing the number of ticks displayed using the clock widget, so it might be nice to also record the number of ticks taken.  Usually this will correspond with the wall clock time, but not if the player explicitly adjusted the tick speed, or if the game slowed down under heavy computational load.

Reply:
> "you tried it but gave up"?

I would not be so pessimistic. :slightly_smiling_face: Three yellow dots are not that bad and I tweaked the update function so it will show the longest time taken while "in progress". So it just means you had "fun" with classic mode for a number of hours. :wink: 

>  So simply using ZonedTime is actually a bit problematic...

Is it? I convert back to UTC when computing elapsed time, so it should be safe AFAIK. :thinking: 

I wanted to keep the time zone in case we wanted to show the time somewhere in-game which we currently do not.

> ...people are going to be used to seeing the number of ticks displayed using the clock widget...

Absolutely! I already added it to the list of tasks in the description, should be easy to add. :+1:

It also makes more sense to measure in ticks if you are loading the solution from a file, then it is just algorithmic complexity and wall clock time does not matter.

Plus the hex clock is cool! :grinning:

Reply:
>  So simply using ZonedTime is actually a bit problematic...

Is it? I convert back to UTC when computing elapsed time, so it should be safe AFAIK. :thinking: 

I wanted to keep the time zone in case we wanted to show the time somewhere in-game which we currently do not.

Reply:
> ...people are going to be used to seeing the number of ticks displayed using the clock widget...

Absolutely! I already added it to the list of tasks in the description, should be easy to add. :+1:

It also makes more sense to measure in ticks if you are loading the solution from a file, then it is just algorithmic complexity and wall clock time does not matter.

Plus the hex clock is cool! :grinning:

Reply:
> Is it? I convert back to UTC when computing elapsed time, so it should be safe AFAIK. thinking
> 
> I wanted to keep the time zone in case we wanted to show the time somewhere in-game which we currently do not.

I think I was not clear.  There is nothing wrong with having a time zone, and indeed, we should have it.  What I meant was that if e.g. you start playing a scenario and the computer clock gets reset for some reason then the elapsed time will be incorrect.  I was thinking of e.g. the switch from standard to summer/daylight saving time, though now that I think of it, that will actually be correctly handled since the time zone will change as well.  The elapsed time can still be incorrect if the computer clock gets reset for some other reason (like if you intentionally set it to a different time) but maybe it's not worth worrying about that.

Once we can save games, then we will also need to worry about keeping track of the time elapsed while actually playing the game, as opposed to just the time between start and completion, but we can worry about that when we get to it.

Reply:
> Three yellow dots are not that bad and I tweaked the update function so it will show the longest time taken while "in progress". So it just means you had "fun" with classic mode for a number of hours.

Sure, I have no problem with showing the longest time you have played, that is fun.  But showing three yellow dots still seems strange to me.  If a scenario has no win condition, then it is not really "in progress", it just means "you have played this scenario before", but that does not seem like a useful thing to show in the UI, and three yellow dots does not seem to communicate the right thing.

Concretely, I would suggest:
- Remove the `o` status indicator for scenarios which have not been started.  We don't really need an icon for this state and all the `o`s just seem like visual noise to me.
- Don't show a status indicator for scenarios with no objectives (but we can still show the longest time played on the right).

Reply:
I still do not see the problem with time zones. The way I understand it is that when I ask for the current time, the computer will give me a precise number. I will also ask for the timezone and store it together with the precise time. This way I can show the time in a meaningful format and perform operations like subtracting an hour, where daylight saving time can complicate things.

But our use case is very simple:
1. get time at the start of the game
2. get time at the end of the game
3. compare them to get elapsed time

The last step should not be affected by timezone gimmicks, because we can convert from `ZonedTime` to `UTCTime` and the `time` library will handle leap seconds for us.

At least that's how I optimistically see it. :smile:

Reply:
Yes, everything you said is true, and I was confused about daylight saving time being a problem.  That and leap seconds will be correctly handled. The only issue (which, to reiterate, is a very small issue and not worth dealing with) is that the computer clock can be changed while the game is running.  Either by e.g. the user explicitly setting the clock, or in some more subtle way e.g. `ntp` adjusting the clock.  So if you say "give me the current ZonedTime" and then later say "give me the current ZonedTime", you are not 100% guaranteed that their difference is the actual amount of elapsed time.

Reply:
Well if we did not trust the players, then saving the high score (and game saves in future) in a human-readable YAML might be a bigger problem. :smile: